airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -263,8 +280,28 @@ def _iter_upstream_conditions() -> Iterator[ColumnOperators]:
             elif trigger_rule == TR.ALL_SKIPPED:
                 if success or failed:
                     new_state = TaskInstanceState.SKIPPED
-
+            elif trigger_rule == TR.ALL_DONE_SETUP_SUCCESS:
+                # when there is an upstream setup and they have all skipped, then skip
+                if upstream_done and upstream_setup and skipped_setup >= upstream_setup:
+                    new_state = TaskInstanceState.SKIPPED
+                    logging.warning(
+                        "ti=%s skipped_setup >= upstream_setup (%s >= %s), marking skipped",
+                        ti.task_id,
+                        skipped_setup,
+                        upstream_setup,
+                    )
+                elif upstream_done and upstream_setup > success_setup:
+                    # when there is an upstream setup, they must all succeed
+                    # otherwise, behave same as all done.
```

Comment:
If it behaves the same as `ALL_DONE` with the exception of AIP-52 related why do we need new trigger rule?
Why can't the existed one be modified to support also AIP-52?

Reply:
so you are saying, if the task is a teardown, it should use ALL_DONE, but, it should behave differently.  so the thing is the behavior isn't "all done"  -- it's all done, _and_ the setup is successful.

Reply:
My question is: teardown are suppose to run even if the task fail thus I don't understand why the user has to specify a new trigger rule?
I may be missing something in the design plan so please correct me if I'm wrong but new introduction of trigger rules specifically for the AIP was not part of the AIP proposal?

Reply:
Oh but so you're saying that new trigger rule was not specified in proposal.  That's correct.  I guess I think of it as implementation detail but, of course such details matter.  The AIP made general predictions about how implementation would proceed, thinking that not setting up deps was going to be better but we found that route problematic.  There's some ongoing discussion of course on the dev list concerning some proposed changes but that's more about API than implementation.  Do you have concerns with adding a new trigger rule to support this?

Reply:
Yeah I have some concerns/questions.

By adding new trigger rules we in fact declare that this is a configuration that a user has to specify manually when he writes the dag. The opens a whole new domain of issues about setting the rule wrong.

Trigger rules are something that many new users get confuse about. I myself answered so many questions related to it in various support channels.

For now I'm missing the part why we must get manual input from the user (trigger rule or other) to get the teardown action working as we want. I always imagined it as if teardown exsist it should run (just like it happes in teardown of test class but again maybe I'm getting this wrong.

Reply:
> When you mark a task as teardown (which is something you must do -- airflow cannot guess about this) that's when it ensures the trigger rule. For teardown tasks, trigger rule is not meant to be configurable

For sure. 
So what you say that we can make it happen by changing the code of current trigger rules and without adding new ones. Thats good.

I just want to point that relaying on trigger rules is problematic. We have cases like `ShortCircuitOperator` that with default execution do NOT honor trigger rules.  
https://github.com/apache/airflow/issues/7858#issuecomment-615863490
Thus from your description if pipeline will use this operator then the teardown might now run

Reply:
OK my goal here was to raise the concern... Now that you are ware of it I think I have nothing else to add so l think It's best that once there is a code/design to review I'll check in again.

As for the ShortCircuitOperator the option to respect trigger rules was already implemented but the operator still offer the option to ignore it.

Reply:
I'll comment on the mailing list thread shortly but in the meantime since this trigger rule is not user facing one but one we set internally in the teardown decorator can we please make it as private? so it will be `_ALL_DONE_SETUP_SUCCESS`

Reply:
Done, switched it to private.